Last Updated : 24AUG2023
# History :
# 24AUG2023 
# 1. Added JointOffSet For each joint. Joint OffSet denotes the value from the corresponding LimitSwitch for the joint
# The value will be in meters for prismatic joint (joint5) and in radians for Joint4 and Joint3
generic_hw_control_loop:
  loop_hz:                    50
#By Mani Radharishnan :changing from 10 to 50
  cycle_time_error_threshold: 0.01

hardware_interface:
  baud: 115200
  odrive_0: "3470368D3031" # This is for Link5 and Link3 Motors
  odrive_1: "347E36623031" #  Odrive_1 is for Link4 and Link2
  joints:
    - joint3
    - joint4
    - joint5
    - joint2   #l2 added by ribin for controlling l2
  sim_control_mode: 0
  odrive_reconfigure_request: false

joint_state_controller:
  type:         joint_state_controller/JointStateController
  publish_rate: 10
#By Mani Radharishnan :changing from 30 to 50

joint3_position_controller:
  type:             position_controllers/JointPositionController
  joint:            joint3
  odrive_id:        0
  axis_id:          1
  CanID:            1
  transmission_factor: 0.870047022 #0.870047022    # Revolutions of motor per radian of ARM=(5.466666667/(2*3.141592654))=0.870047022 , #where 5.466666 is Gear Ratio
  resolution:       1
  direction:        -1
  p_gain:           35
  v_gain:           0.000549
  v_int_gain:       0.000549
  max_t:            80
  max_e:            100
  max_cur:          40
  max_vel:          50
  min_vel:          27
  reference_trgt:  -6.28318     # in radians, towards limit switch
  homing_pos:       0.00 #(1.21 previously-now change by gokul )1.51844 is equal to 87 degree from Limit Switch  #-0.872664626 # 20 deg from Horizontal
  homing_pos_init_seq_: 0.1745  #1.5708
  JointOffSet:       -0.1745 #previusly (1.3845) # this comes from the Design Spec, value in Radian from LimitSwitch -changed by gokul
  switch:           /joint3_homing_service
  limit_switch:     26

joint4_position_controller:
  type:             position_controllers/JointPositionController
  joint:            joint4
  odrive_id:        1
  axis_id:          0
  CanID :           2
  transmission_factor: 14.3164  #12.7325  #0.870047022 # Revolutions of motor per radian of ARM=(5.466666667/(2*3.141592654))=0.870047022 ,
                                    #where 5.466666 is Gear Ratio
  resolution:       1
  direction:        1
  p_gain:           100
  v_gain:           0.000928
  v_int_gain:       0.000928
  max_t:            80
  max_e:            100
  max_cur:          50
  max_vel:          60
  min_vel:          27
  reference_trgt:   6.28318
  homing_pos:       0.000    #-2.65-(previously for rotation )  #2.35 #3.34  #4.71238898 #1.15192 #0.627993  #1.15192  #2.72271 #1.57 rad from limit switch hit  90 degree is 1.570 radians , 2.057781radians is  117.9 deg TODO change this >
  JointOffSet      : 0.000 #-2.65-(previously for rotation )  # this comes from the Design Spec, value in Radian from LimitSwitch
  switch:           /joint4_homing_service
  limit_switch:     20

joint5_position_controller:
  type:             position_controllers/JointPositionController
  joint:            joint5
  odrive_id:        0
  axis_id:          0
  CanID:            0
  transmission_factor: 12.7325
 # Revolutions of motor per meter of movement ARM=(5.466666667รท0.15)=36.4444 (for testing dec_9 1/0.0785=12.7325 by gokul)
                               # where 5.46666667 is Gear ratio and 0.24 is lead of the screw
  resolution:       1
  direction:        -1
  p_gain:           35
  v_gain:           0.000549
  v_int_gain:       0.000549
  max_t:            80
  max_e:            100
  max_cur:          40
  max_vel:          133
  min_vel:          27
  reference_trgt:   -6.28318
  homing_pos:       0.005 # 0.247 + 0.005 5mm away from the limit switch
  JointOffSet      : -0.213 #(-0.247) this comes from the Design Spec, value in meters from LimitSwitch
  switch:           /joint5_homing_service
  limit_switch:     16

##################################added by ribin for l2 control################################
joint2_position_controller:
    type:             position_controllers/JointPositionController
    joint:            joint2
    odrive_id:        1
    axis_id:          1
    CanID:            3
    transmission_factor: 125.23664 #78.2729 # Revolutions of motor per meter of movement ARM=(5.475/0.24)=22.8125
                                 # where 5.475 is Gear ratio and 0.24 is lead of the screw
    resolution:       1
    direction:        1
    p_gain:           35
    v_gain:           0.000549
    v_int_gain:       0.000549
    max_t:            80
    max_e:            100
    max_cur:          20
    max_vel:          133
    min_vel:          27
    reference_trgt:   -6.28318
    homing_pos:       0.03
    JointOffSet      : 0.03 # this comes from the Design Spec, value in meters from LimitSwitch
    switch:           /joint2_homing_service
    limit_switch:     5

###################################################l2 control#############################
joint_position_controller:
  type:  position_controllers/JointGroupPositionController
  joints:
    - joint3
    - joint4
    - joint5
    - joint2   #added by ribin for l2 control

